### ISSUE TITLE
Compiler gives field-never-assigned warning for private struct

### ISSUE THREAD

**Issue by Inverness** [dotnet/roslyn#2470]
In VS2015 RC, the compiler will give you an unassigned field warning 649 for a private struct. This seems misleading because its quite possible for a struct to be properly used when initialized to its default value.

I think this warning should only apply for reference types and primitive value types that are immutable.

**Comment by gafter** [comment]
Can you please provide a small sample program that demonstrates the scenario you're encountering?

**Comment by Inverness** [comment]
Sure. I encounter the warning when using a helper struct, designed to be used as a private field, to help with implementing asynchronous events.

```C#
    internal struct TaskEvent<T>
    {
        private TaskCompletionSource<T> _tcs;

        public Task<T> Task
        {
            get
            {
                if (_tcs == null)
                    _tcs = new TaskCompletionSource<T>();
                return _tcs.Task;
            }
        }

        public void Invoke(T result)
        {
            if (_tcs != null)
            {
                TaskCompletionSource<T> localTcs = _tcs;
                _tcs = null;
                localTcs.SetResult(result);
            }
        }
    }

    public class OperationExecutor
    {
        private TaskEvent<float?> _nextValueEvent; // Field is never assigned warning

        // Start some async operation
        public Task<bool> StartOperation()
        {
            return ...;
        }

        // Get progress or data during async operation
        public Task<float?> WaitNextValue()
        {
            return _nextValueEvent.Task;
        }

        // Called externally
        internal void OnNextValue(float? value)
        {
            _nextValueEvent.Invoke(value);
        }
    }
```

**Comment by jaredpar** [comment]
It looks like this bug has been fixed in RTM.  Need to add a unit test to ensure this doesn't get regressed.
